src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx

```diff
@@ -63,15 +98,412 @@ const SicpNavigationBar: React.FC = () => {
+    hasBackdrop: true,
+    isOpen: isSearchOpen,
+    position: Position.LEFT,
+    usePortal: false
+  };
+
+  function queryIndexTrie(query: string) {
+    let node = indexTrie;
+    for (let i = 0; i < query.length; i++) {
+      const char = query[i];
+      if (node[char]) {
+        node = node[char];
+      } else {
+        return [];
+      }
+    }
+    if (node['value']) {
+      return node['value'];
+    }
+    return [];
+  }
+
+  function queryTextTrie(query: string) {
+    let node = textTrie;
+    for (let i = 0; i < query.length; i++) {
+      const char = query[i];
+      if (node[char]) {
+        node = node[char];
+      } else {
+        return [];
+      }
+    }
+    if (node['value']) {
+      return node['value'];
+    }
+    return [];
+  }
+
+  function autoComplete(str: string, limit: number, jsonData: any) {
+    if (str.length === 0) {
+      return [];
+    }
+    str = str.toLowerCase();
+    const trie = jsonData;
+
+    function next(node: any, i: number) {
+      if (i === 26) return node[' '];
+      else if (i >= 0 && i <= 25) {
+        return node[String.fromCharCode(i + 97)];
+      }
+      return null;
+    }
+    const letters = [
+      'a',
+      'b',
+      'c',
+      'd',
+      'e',
+      'f',
+      'g',
+      'h',
+      'i',
+      'j',
+      'k',
+      'l',
+      'm',
+      'n',
+      'o',
+      'p',
+      'q',
+      'r',
+      's',
+      't',
+      'u',
+      'v',
+      'w',
+      'x',
+      'y',
+      'z',
+      ' '
+    ];
+
+    function toEnd(query: any) {
+      let node = trie;
+      for (let i = 0; i < query.length; i++) {
+        const char = query[i];
+        if (node[char]) {
+          node = node[char];
+        } else {
+          return null;
+        }
+      }
+      return node;
+    }
+
+    function recur(node: any, path: string) {
+      if (node == null || limit < 0) {
+        return;
+      }
+      if (node.value) {
+        ans.push(path);
+        limit--;
+      }
+      for (let i = 0; i < 27; i++) {
+        recur(next(node, i), path + letters[i]);
+      }
+    }
+    const ans = [] as any[];
+    recur(toEnd(str), str);
+    return ans;
+  }
+
+  const voidSearch = (str: string) => {
+    function lineMap(array: any[]) {
+      if (array == null || array[0] == null || array[1] == null || array[2] == null) {
+        return { title: '', url: '' };
+      }
+      return {
+        title: textbook[array[0]][array[1]][array[2]],
+        url: SearchUrl + array[0] + array[1]
+      };
+    }
+    const SearchUrl = '..';
+    let words = str.toLowerCase().split(' ');
+    words = words.filter(word => word !== '');
+    if (searchQuery.length === 0 || words.length === 0) {
+      setQueryResult([]);
+      return;
+    }
+    let globalAns = queryTextTrie(words[0]).map((array: any) => lineMap(array));
+
+    globalAns = globalAns.filter((obj: any) => obj.title.toLowerCase().includes(str.toLowerCase()));
+    return globalAns;
```

Comment:
Let us prefer to use `const`. The mutation here is unnecessary:

```suggestion
    const globalAns = queryTextTrie(words[0]).map((array: any) => lineMap(array));
    return globalAns.filter((obj: any) => obj.title.toLowerCase().includes(str.toLowerCase()));
```

Also note the use of the unaliased `any` type (not addressed in the suggestion above).

Reply:
I've fixed this in #2471.